File: cpp-answers/144.二叉树的前序遍历.cpp
/*
 * @Description: 
 * @Author: Hongyang_Yang
 * @Date: 2020-08-19 10:10:03
 * @LastEditors: Hongyang_Yang
 * @LastEditTime: 2020-08-19 10:19:14
 */
/*
 * @lc app=leetcode.cn id=144 lang=cpp
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    /*
    void preorder(TreeNode *root, vector<int> &res)
    {
        if (!root)
            return;
        res.push_back(root->val);
        preorder(root->left, res);
        preorder(root->right, res);
    }
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> res;
        preorder(root, res);
        return res;
    }
    */
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> res;
        stack<TreeNode *> st;
        if (root == NULL)
            return res;
        st.push(root);
        while (!st.empty())
        {
            TreeNode *tmp = st.top();
            st.pop();
            res.push_back(tmp->val);
            if (tmp->right)
                st.push(tmp->right);
            if (tmp->left)
                st.push(tmp->left);
        }
        return res;
    }
};
// @lc code=end

File: rust-answers/144.二叉树的前序遍历.rs
/*
 * @lc app=leetcode.cn id=144 lang=rust
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

type NodeOpt = Option<Rc<RefCell<TreeNode>>>;

impl Solution {
    pub fn preorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        Self::preorder_traversal_iter(root)

        //let mut result = vec![];
        // Self::preorder_traversal_recurs(root, &mut result);
        // result
    }

    fn preorder_traversal_iter(root: NodeOpt) -> Vec<i32> {
        let mut result = vec![];
        let mut stack = vec![root];
        while let Some(node_opt) = stack.pop() {
            if let Some(pnode) = node_opt {
                let node = pnode.borrow();
                result.push(node.val);
                stack.push(node.right.clone());
                stack.push(node.left.clone());
            }
        }
        result
    }
}
// @lc code=end

